src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs

```diff
@@ -0,0 +1,813 @@
+                                                                .ToImmutableArray();
+            if (originalDeclStatementsToMove.IsEmpty)
+            {
+                return memberDeclaration;
+            }
+
+            // Moving declarations closer to a reference can lead to conflicting edits.
+            // So, we track all the declaration statements to be moved upfront, and update
+            // the root, document, editor and memberDeclaration for every edit.
+            // Finally, we apply replace the memberDeclaration in the originalEditor as a single edit.
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var rootWithTrackedNodes = root.TrackNodes(originalDeclStatementsToMove);
+
+            // Run formatter prior to invoking IMoveDeclarationNearReferenceService.
+            rootWithTrackedNodes = Formatter.Format(rootWithTrackedNodes, originalDeclStatementsToMove.Select(s => s.Span), document.Project.Solution.Workspace, cancellationToken: cancellationToken);
+
+            document = document.WithSyntaxRoot(rootWithTrackedNodes);
+            await OnDocumentUpdatedAsync().ConfigureAwait(false);
+
+            foreach (TLocalDeclarationStatementSyntax originalDeclStatement in originalDeclStatementsToMove)
+            {
+                // Get the current declaration statement.
+                var declStatement = memberDeclaration.GetCurrentNode(originalDeclStatement);
+
+                var documentUpdated = false;
+
+                // Check if the new variable declaration is unused after all the fixes, and hence can be removed.
+                if (await TryRemoveUnusedLocalAsync(declStatement).ConfigureAwait(false))
+                {
+                    documentUpdated = true;
+                }
+                else
+                {
+                    // Otherwise, move the declaration closer to the first reference if possible.
+                    if (await service.CanMoveDeclarationNearReferenceAsync(document, declStatement, cancellationToken).ConfigureAwait(false))
+                    {
+                        document = await service.MoveDeclarationNearReferenceAsync(document, declStatement, cancellationToken).ConfigureAwait(false);
+                        documentUpdated = true;
+                    }
+                }
+
+                if (documentUpdated)
+                {
+                    await OnDocumentUpdatedAsync().ConfigureAwait(false);
+                }
+            }
+
+            return memberDeclaration;
+
+            // Local functions.
+            async Task OnDocumentUpdatedAsync()
+            {
+                root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                memberDeclaration = syntaxFacts.GetContainingMemberDeclaration(root, memberDeclaration.SpanStart);
```

Comment:
`MoveNewLocalDeclarationsNearReferenceAsync` only works on nodes within the member declaration, so the member declaration's span start cannot change within a single invocation to `MoveNewLocalDeclarationsNearReferenceAsync`. Implementation of `PostProcessDocumentCoreAsync` ensures this is safe across invocations to `MoveNewLocalDeclarationsNearReferenceAsync` for different method bodies in the same source file:

```
                foreach (var memberDecl in newRoot.DescendantNodes().Where(n => n.HasAnnotation(s_memberAnnotation)))
                {
                    var newMemberDecl = await processMemberDeclarationAsync(memberDecl, newDocument, cancellationToken).ConfigureAwait(false);
                    memberDeclReplacementsMap.Add(memberDecl, newMemberDecl);
                }

                return newRoot.ReplaceNodes(memberDeclReplacementsMap.Keys,
                    computeReplacementNode: (node, _) => memberDeclReplacementsMap[node]);
```

---
In reply to: [236088685](https://github.com/dotnet/roslyn/pull/30777#discussion_r236088685) [](ancestors = 236088685)

Reply:
> 2. the way these local functions mutate local state is super unclear.  can we not have this local function return the member decl, and have the caller assign over the local?  

I personally don't see any readability issue with this. I see the local functions mutating local state as equivalent to private instance methods in helper types, such that that local state is captured into fields in the constructor of the helper type, there is a single public entry point into the helper type which has the local function's containing method implementation.
Regardless, I feel the only readable implementation with non state mutating local functions would require static local functions, so I am leaving this as si for now.

---
In reply to: [236347484](https://github.com/dotnet/roslyn/pull/30777#discussion_r236347484) [](ancestors = 236347484,236088685)

Reply:
2. THe difference is that with class state you know the only way to mutate is through methods.  With local state that's rarely the case.  99% of mutations of local state is through local flow.  Local functions came recently and now act differently than all otehr code we've had.  Note that you could mutate local state with lambdas.  But this was also a 0.00001% case.

So, local function mutation is obviously semantically fine.  However, it adds a lot of cognitive burden because these methods now behave unlike almost every other line of code in the IDE.

Reply:
1. 

> MoveNewLocalDeclarationsNearReferenceAsync only works on nodes within the member declaration, so the member declaration's span start cannot change within a single invocation to MoveNewLocalDeclarationsNearReferenceAsync. Implementation of 

This does feel a bit brittle.  But potentially worth documenting.  For example, if this is in a separate service, it's not always clear that the above will stay true.